### ISSUE TITLE
Writing data object via ticket => CAT_SQL_ERR, inconsistent DATA_SIZE

### ISSUE THREAD

**Issue by cjsmeele** [irods/irods#4744]
- [x] master
- [x] 4-2-stable
------

### iRODS Version, OS and Version

4.2.7, Ubuntu 16.04

### What did you try to do?

Use `iput -t TICKET` to overwrite a data object.

Originally using open/write/close API calls, but this problem also occurs with regular icommands.

### Expected behavior

No errors occur. File is overwritten.

### Observed behavior (including steps to reproduce, if applicable)

When the new file has the same size, there are no issues - the file is overwritten as expected.

However:

#### When new size is larger

Open & write calls succeed, down to the resource level AFAIU (`pep_resource_write_post` fires before any error occurs).

- Put fails with a `CAT_SQL_ERR` (see log below).
- icat data size is unchanged
- reading the file (`iget`) returns the new file contents, truncated to the old size

#### When new size is smaller

Same as above, but reading the file returns the new contents, shorter than the size reported by `ils` (!)

#### Reproducing

```
% iuserinfo | grep type
type: rodsuser
% printf foo123 > test.txt
% iput test.txt
% iticket create write test.txt TICK
% iget -t TICK test.txt - | xxd
00000000: 666f 6f31 3233                           foo123
% 
% # Overwrite with smaller file.
% 
% printf bar > test.txt
% iput -f -t TICK test.txt test.txt
remote addresses: 127.0.0.1 ERROR: putUtil: put error for /tempZone/home/chris/test.txt, status = -806000 status = -806000 CAT_SQL_ERR
% ils -L test.txt
  chris             0 demoResc            6 2020-02-25.13:06 & test.txt
        generic    /var/lib/irods/Vault/home/chris/test.txt
% iget -t TICK test.txt - | xxd
00000000: 6261 72                                  bar
% xxd /var/lib/irods/Vault/home/chris/test.txt
00000000: 6261 72                                  bar
% 
% # Overwrite with larger file
% 
% printf barbar123 > test.txt
% iput -f -t TICK test.txt test.txt
remote addresses: 127.0.0.1 ERROR: putUtil: put error for /tempZone/home/chris/test.txt, status = -806000 status = -806000 CAT_SQL_ERR
% ils -L test.txt
  chris             0 demoResc            6 2020-02-25.13:06 & test.txt
        generic    /var/lib/irods/Vault/home/chris/test.txt
% iget -t TICK test.txt - | xxd
00000000: 6261 7262 6172                           barbar
% xxd /var/lib/irods/Vault/home/chris/test.txt
00000000: 6261 7262 6172 3132 33                   barbar123
% 
```

### Short analysis

The rodslog shows that the failing query has to do with the `write_byte_count` of the session ticket. Because of #2719, this query is likely not executed in the happy path above, where the written size equals the existing size ([1](https://github.com/irods/irods/blob/4.2.7/plugins/database/src/db_plugin.cpp#L2245), [2](https://github.com/irods/irods/blob/4.2.7/plugins/database/src/db_plugin.cpp#L2351-L2354)).

The query appears to be broken [here](https://github.com/irods/irods/commit/6ed2c6fd3e3ad6ce70e57371c8952e8ff4aab976#diff-233f93898a37daf33d04fd50fb63f47eR1599-R1604) (changed bindvars with unchanged query).
I hope this helps!

### Log

The following was printed to the rodsLog during `iput -f -t ...`, overwriting test.txt with a smaller file.

```
Feb 25 13:25:59 pid:6043 NOTICE: session ticket setup, value: TICK
Feb 25 13:25:59 pid:6043 NOTICE: bindVar1=TICK
Feb 25 13:25:59 pid:6043 NOTICE: bindVar2=10024
Feb 25 13:25:59 pid:6043 NOTICE: bindVar3=/tempZone/home/chris
Feb 25 13:25:59 pid:6043 NOTICE: cllExecSqlWithResultBV: SQLExecDirect error: -1, sql:select ticket_id, write_byte_count, write_byte_limit from R_TICKET_MAIN where ticket_type = 'write' and ticket_string = ? and (object_id = ? or object_id in (select coll_id from R_DATA_MAIN where data_id = ?))
Feb 25 13:25:59 pid:6043 NOTICE: SQLSTATE: 22P02
Feb 25 13:25:59 pid:6043 NOTICE: SQLCODE: 7
Feb 25 13:25:59 pid:6043 NOTICE: SQL Error message: ERROR: invalid input syntax for integer: "/tempZone/home/chris";
Error while executing the query
Feb 25 13:25:59 pid:6043 NOTICE: bindVar1=TICK
Feb 25 13:25:59 pid:6043 NOTICE: bindVar2=10024
Feb 25 13:25:59 pid:6043 NOTICE: bindVar3=/tempZone/home
Feb 25 13:25:59 pid:6043 NOTICE: cllExecSqlWithResultBV: SQLExecDirect error: -1, sql:select ticket_id, write_byte_count, write_byte_limit from R_TICKET_MAIN where ticket_type = 'write' and ticket_string = ? and (object_id = ? or object_id in (select coll_id from R_DATA_MAIN where data_id = ?))
Feb 25 13:25:59 pid:6043 NOTICE: SQLSTATE: 22P02
Feb 25 13:25:59 pid:6043 NOTICE: SQLCODE: 7
Feb 25 13:25:59 pid:6043 NOTICE: SQL Error message: ERROR: invalid input syntax for integer: "/tempZone/home";
Error while executing the query
Feb 25 13:25:59 pid:6043 NOTICE: bindVar1=TICK
Feb 25 13:25:59 pid:6043 NOTICE: bindVar2=10024
Feb 25 13:25:59 pid:6043 NOTICE: bindVar3=/tempZone
Feb 25 13:25:59 pid:6043 NOTICE: cllExecSqlWithResultBV: SQLExecDirect error: -1, sql:select ticket_id, write_byte_count, write_byte_limit from R_TICKET_MAIN where ticket_type = 'write' and ticket_string = ? and (object_id = ? or object_id in (select coll_id from R_DATA_MAIN where data_id = ?))
Feb 25 13:25:59 pid:6043 NOTICE: SQLSTATE: 22P02
Feb 25 13:25:59 pid:6043 NOTICE: SQLCODE: 7
Feb 25 13:25:59 pid:6043 NOTICE: SQL Error message: ERROR: invalid input syntax for integer: "/tempZone";
Error while executing the query
Feb 25 13:25:59 pid:6043 remote addresses: 127.0.0.1, ::1 ERROR: [-]	/irods/server/api/src/rsModDataObjMeta.cpp:164:int _rsModDataObjMeta(rsComm_t *, modDataObjMeta_t *) :  status [CAT_SQL_ERR]  errno [] -- message [_rsModDataObjMeta - Failed to modify the database for object "/tempZone/home/chris/test.txt" - CAT_SQL_ERR ]

Feb 25 13:25:59 pid:6043 NOTICE: _rsModDataObjMeta - Failed updating the database with object info.
Feb 25 13:25:59 pid:6043 NOTICE: Warning, pending SQL at cllDisconnect, count: 2
Feb 25 13:25:59 pid:6043 NOTICE: Warning, pending SQL: begin ...
Feb 25 13:25:59 pid:6043 NOTICE: Warning, pending SQL: update R_TICKET_MAIN set writ ...
Feb 25 13:25:59 pid:6043 NOTICE: Warning, SQL will be commited with audits
Feb 25 13:25:59 pid:6043 NOTICE: Notice, pending Auditing SQL committed at cllDisconnect
```

**Comment by korydraughn** [comment]
Verified that the CAT_SQL_ERR has been resolved.  The only thing left for this is to update the catalog to reflect the correct size on disk.

**Comment by korydraughn** [comment]
This has been resolved in master and 4-2-stable. Running the script against the latest branches produces the following output:
```bash
+ iuserinfo
+ grep type
type: rodsadmin
+ printf foo123
+ iput test.txt
+ iticket create write test.txt TICK
+ iget -t TICK test.txt -
+ xxd
00000000: 666f 6f31 3233                           foo123
+ printf bar
+ iput -ft TICK test.txt test.txt
+ ils -L test.txt
  kory              0 demoResc            3 2021-01-28.09:52 & test.txt
        generic    /var/lib/irods/Vault/home/kory/test.txt
+ iget -t TICK test.txt -
+ xxd
00000000: 6261 72                                  bar
+ sudo xxd /var/lib/irods/Vault/home/kory/test.txt
00000000: 6261 72                                  bar
+ printf barbar123
+ iput -ft TICK test.txt test.txt
+ ils -L test.txt
  kory              0 demoResc            9 2021-01-28.09:52 & test.txt
        generic    /var/lib/irods/Vault/home/kory/test.txt
+ iget -t TICK test.txt -
+ xxd
00000000: 6261 7262 6172 3132 33                   barbar123
+ sudo xxd /var/lib/irods/Vault/home/kory/test.txt
00000000: 6261 7262 6172 3132 33                   barbar123
```

**Comment by trel** [comment]
Note the linked issue https://github.com/irods/irods/issues/2719 as well.

**Comment by trel** [comment]
Please note when this was resolved since no commit has referenced this issue directly.

**Comment by korydraughn** [comment]
This issue was resolved by https://github.com/irods/irods/commit/d91ec25e619e159c5b9416c7fed2408d31dbf833. I have verified that the script included in this issue results in bad catalog data when https://github.com/irods/irods/commit/d91ec25e619e159c5b9416c7fed2408d31dbf833 is not included. When https://github.com/irods/irods/commit/d91ec25e619e159c5b9416c7fed2408d31dbf833 is included, the catalog reflects the correct information.

Therefore, this issue can be closed.
